build_tools/cmake/iree_run_module_test.cmake

```diff
@@ -0,0 +1,369 @@
+#   EXPECTED_OUTPUT: A file of expected output to compare with the output from
+#       iree-run-module
+#   LABELS: Additional labels to apply to the test. The package path and
+#       "driver=${DRIVER}" are added automatically.
+#   XFAIL: When set for x86_64|android-arm64-v8a|riscv64-Linux|riscv32-Linux,
+#       the test is expected to fail for the particular platforms/architectures
+#       for various reasons, e.g., upstream llvm backend.
+#   UNSUPPORTED_PLATFORMS: Platforms
+#       (android-arm64-v8a|riscv64-Linux|riscv32-Linux) not supported by the
+#       test target. The target will be skipped during crosscompile.
+#   DEPS: (Optional) List of targets to build the test artifacts.
+#   TIMEOUT: (optional) Test timeout.
+#
+# Examples:
+#
+# iree_run_module_test(
+#   NAME
+#     iree_run_module_correctness_test
+#   MODULE_SRC
+#     "iree_run_module_bytecode_module_llvm_cpu.vmfb"
+#   DRIVER
+#     "local-sync"
+#   RUNNER_ARGS
+#     "--entry_function=abs"
+#     "--function_input=f32=-10"
+#   EXPECTED_OUTPUT
+#     "f32=10"
+# )
+#
+# iree_run_module_test(
+#   NAME
+#     person_detect_int8_correctness_test
+#   MODULE_SRC
+#     "person_detect_int8.vmfb"
+#   DRIVER
+#     "local-sync"
+#   RUNNER_ARGS
+#     "--entry_function=main"
+#     "--function_input=1x96x96x1xi8=0"
+#   EXPECTED_OUTPUT
+#     "1x2xi8=[72 -72]"
+#   UNSUPPORTED_PLATFORMS
+#     "android-arm64-v8a"
+# )
+#
+# iree_run_module_test(
+#   NAME
+#     mobilenet_v1_fp32_correctness_test
+#   MODULE_SRC
+#     "mobilenet_v1_fp32.vmfb"
+#   DRIVER
+#     "local-sync"
+#   RUNNER_ARGS
+#     "--entry_function=main"
+#     "--function_input=1x224x224x3xf32=0"
+#   EXPECTED_OUTPUT
+#     "mobilenet_v1_fp32_expected_output.txt"
+#   UNSUPPORTED_PLATFORMS
+#     "android-arm64-v8a"
+#     "riscv32-Linux"
+# )
+
+function(iree_run_module_test)
+  if(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME;MODULE_SRC;DRIVER;EXPECTED_OUTPUT;TIMEOUT"
+    "RUNNER_ARGS;LABELS;XFAIL;UNSUPPORTED_PLATFORMS;DEPS"
+    ${ARGN}
+  )
+
+  if(CMAKE_CROSSCOMPILING AND "hostonly" IN_LIST _RULE_LABELS)
+    return()
+  endif()
+
+  iree_get_platform(_PLATFORM)
+  if(_PLATFORM IN_LIST _RULE_UNSUPPORTED_PLATFORMS)
+    return()
+  endif()
+
+  set(_RUNNER_TARGET "iree-run-module")
+  iree_package_name(_PACKAGE_NAME)
+  set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+  get_filename_component(_SRC "${_RULE_MODULE_SRC}" REALPATH)
+  list(APPEND _RULE_RUNNER_ARGS "--module_file=${_SRC}")
+
+  if(_RULE_EXPECTED_OUTPUT)
+    get_filename_component(_OUTPUT_FILE_TYPE "${_RULE_EXPECTED_OUTPUT}" EXT)
+    if(_OUTPUT_FILE_TYPE STREQUAL ".txt")
+      get_filename_component(_OUTPUT_FILE_SRC "${_RULE_EXPECTED_OUTPUT}" REALPATH)
+      # Process the text input to remove the line breaks.
+      file(STRINGS "${_OUTPUT_FILE_SRC}" _EXPECTED_OUTPUT ENCODING UTF-8)
+      string(REPLACE ";" " " _EXPECTED_OUTPUT_STR "${_EXPECTED_OUTPUT}")
```

Comment:
The `file(STRINGS ...)` `list(JOIN ...` combo can't remove the line breaks in the text file to generate the correct `--expected_output=` one-liner. Keep the original call.

Reply:
Huh, really? I'd expect this to do exactly the same thing? OOC what does it generate?

Reply:
it's the literal string of the text file, i.e.,
```
--expected_output="1x1000xi8=[
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2
0 0 1 0 0 0 0 0 0 3 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 7 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0
1 0 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0
0 0 0 1 0 0 0 1 0 0 3 0 0 0 3 0 0 0 0 1 0 0 1 1 0 0 0 0 0 4 0 0 0 0 0 0 3 0 0 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 2 1 1 0 0 0 0 0 0 1 0 0 1
2 0 0 0 1 1 1 0 0 0 0 0 0 0 0 1 1 0 1 1 0 0 0 1 0 0 1 0 0 1 0 0 6 1 0 0 1 0 0 0
0 0 0 0 4 0 0 0 0 0 2 5 0 0 0 0 0 0 0 0 0 0 1 0 3 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0
0 1 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0
0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 5 1 0 0 0 0 1
1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 0 0 0 0 5 1 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0
0 0 0 0 1 0 0 0 0 0 0 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 1 0 1 0 0 0 0 5 1 0 5 0 0 0 0 1 0 1 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 1 0 0
0 0 0 0 0 2 0 0 0 0 0 0 1 2 0 0 1 0 0 0 0 0 1 0 3 2 1 0 0 0 0 0 0 0 0 0 2 0 0 0
1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]"
```
while `iree-run-module` expects
```
--expected_output="1x1000xi8=[ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 1 0 0 0 0 0 0 3 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 7 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 3 0 0 0 3 0 0 0 0 1 0 0 1 1 0 0 0 0 0 4 0 0 0 0 0 0 3 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 2 1 1 0 0 0 0 0 0 1 0 0 1 2 0 0 0 1 1 1 0 0 0 0 0 0 0 0 1 1 0 1 1 0 0 0 1 0 0 1 0 0 1 0 0 6 1 0 0 1 0 0 0 0 0 0 0 4 0 0 0 0 0 2 5 0 0 0 0 0 0 0 0 0 0 1 0 3 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 0 1 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 5 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 0 0 0 0 5 1 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 5 1 0 5 0 0 0 0 1 0 1 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 2 0 0 0 0 0 0 1 2 0 0 1 0 0 0 0 0 1 0 3 2 1 0 0 0 0 0 0 0 0 0 2 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]"
```

Reply:
That didn't make sense to me, so I tested it with a simple test file. `STRINGS` + `JOIN` behaves as I expected and gives a single line.

```cmake
# test_join.cmake
cmake_minimum_required(VERSION 3.21...3.23)

set(_BASENAME "test_file")
set(_INPUT_FILE "${_BASENAME}.txt")
set(_JOIN_OUTPUT_FILE "${_BASENAME}.join.txt")
set(_REPLACE_OUTPUT_FILE "${_BASENAME}.replace.txt")

file(STRINGS "${_INPUT_FILE}" _CONTENT ENCODING UTF-8)

list(LENGTH _CONTENT _LENGTH)

message(STATUS "${_LENGTH}")
message(STATUS "${_CONTENT}")

string(REPLACE ";" " " _REPLACE_OUTPUT "${_CONTENT}")
list(JOIN _CONTENT " " _JOIN_OUTPUT)
file(WRITE "${_JOIN_OUTPUT_FILE}" "${_JOIN_OUTPUT}")
file(WRITE "${_REPLACE_OUTPUT_FILE}" "${_REPLACE_OUTPUT}")

execute_process(COMMAND ${CMAKE_COMMAND} -E compare_files "${_JOIN_OUTPUT_FILE}" "${_REPLACE_OUTPUT_FILE}"
                RESULT_VARIABLE compare_result
)

if( compare_result EQUAL 0)
    message(STATUS "The files are identical.")
elseif( compare_result EQUAL 1)
    message(SEND_ERROR "The files are different.")
else()
    message(SEND_ERROR "Error while comparing the files.")
endif()
```

```shell
$ cmake -P test_join.cmake 
-- 3
-- foo;bar;baz
-- The files are identical.
```

But if I switch to using "efficientnet_int8_expected_output.txt" wget from your change, I get different results:

```
$ cmake -P test_join.cmake 
-- 1
-- 1x1000xi8=[;0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 0 0 0 0 0 0 0 0;0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
CMake Error at test_join.cmake:27 (message):
  The files are different.
```

Somehow `file(STRINGS ...)` is giving a single string with semicolons in it, which CMake *should* be interpreting as a list, but it isn't for some reason...

I found this bug: https://public.kitware.com/Bug/view.php?id=9317. It turns out that CMake is just broken with brackets in a list. That bug tracker and the bug were abandoned. The closest I can find in the new bug tracker is a request to "document" this "feature" (https://gitlab.kitware.com/cmake/cmake/-/issues/21808) and then this discourse discussion (https://discourse.cmake.org/t/why-are-square-brackets-special-in-lists-problem-with-cmake-parse-arguments/1229) confirming that this will be broken forever. WTF CMake...

So basically lists with square brackets in them are impossible. I'd suggest avoiding them entirely and not using `file(STRINGS ...)` here. How about 

```cmake
      file(READ "${_OUTPUT_FILE_SRC}" _EXPECTED_OUTPUT)
      string(REPLACE "\n" " " _EXPECTED_OUTPUT "${_EXPECTED_OUTPUT}")
```

This results in a slightly different file, but I think functionally the same. Somehow the other one ends up with an extra EOF or something:

```shell
$ cmp -b efficientnet_int8_expected_output.raw.txt efficientnet_int8_expected_output.strings.txt 
cmp: EOF on efficientnet_int8_expected_output.strings.txt after byte 2012, in line 1
```

And then probably we should avoid using CMake for doing file processing at all. Reading content into CMake appears to be broken in a number of different ways

Reply:
Update: I was misinterpreting the `cmp` output. The file generated by `READ` + `REPLACE` has a trailing space. That should be fine.